### ISSUE TITLE
serialize_u64 is slow

### ISSUE THREAD

**Issue by dtolnay** [serde-rs/json#84]
It currently looks like this:

```rust
fn serialize_u64(&mut self, value: u64) -> Result<()> {
    write!(&mut self.writer, "{}", value).map_err(From::from)
}
```

I tested the following variation and it made an astonishing 20% difference in the [bench_log](https://github.com/serde-rs/json/blob/57739801bf53793062c84c5f144d07ce2f126a5d/json_tests/benches/bench_log.rs) benchmarks (294 MB/s vs 244 MB/s).

```rust
if value == 0 {
    self.writer.write_all(b"0").map_err(From::from)
} else {
    let mut scratch = [0u8; 20];
    let mut len = 0;
    while value > 0 {
        scratch[19 - len] = (value % 10) as u8 + b'0';
        value /= 10;
        len += 1;
    }
    self.writer.write_all(&scratch[20-len..20]).map_err(From::from)
}
```

That is enough to motivate looking for a better solution, or contributing a fix to Rust for whatever is making `write!` slow.

**Comment by dtolnay** [comment]
Here is the code that `write!` is using: [num.rs#L201-L254](https://github.com/rust-lang/rust/blob/c128e9bb2e58d7e2fe9679ac81c33a859522d5a5/src/libcore/fmt/num.rs#L201-L254).

**Comment by dtolnay** [comment]
The `fmt` method I linked to is way faster than mine. All the slowness comes from [`pad_integral`](https://github.com/rust-lang/rust/blob/c128e9bb2e58d7e2fe9679ac81c33a859522d5a5/src/libcore/fmt/mod.rs#L906-L966) which we shouldn't even need because we want the default format. I will look for a way to avoid going through `pad_integral`.

**Comment by dtolnay** [comment]
Fixed in #87.
